### ISSUE TITLE
Memory leak when plotting dynamic data

### ISSUE THREAD

**Issue by daltzctr** [beto-rodriguez/LiveCharts2#1080]
**Describe the bug**
I have an application that is continuously plotting ~200/s. Plotting performance itself is acceptable, but the plot leaks memory and eventually performance halts (stuttering as the leak expands). It looks small, as I don't notice it in normal use (30 minute bursts), but after doing some stress tests, I was able to reproduce 1.5G of memory used after ~12 hours.

![image](https://github.com/beto-rodriguez/LiveCharts2/assets/105223895/1d3328fe-e194-49af-823a-e753580df77d)

**To Reproduce**
1. Add a line of points continuously to a LiveCharts2 plot
2. Run overnight, come back next day

**Expected behavior**
No memory leaks, consistent performance in long scenarios

**Screenshots**
Already attached

**Desktop (please complete the following information):**
 - OS: Windows 11 22621.1848
 - Browser: NA
 - Version: v2.0.0-beta.800
 - .NET Version: .NET 8 Preview 5
 
**Additional context**
Used in a MAUI WinUI3 project

**Comment by daltzctr** [comment]
dotMemory analysis 
![image](https://github.com/beto-rodriguez/LiveCharts2/assets/105223895/bd25a6a8-0d38-43c3-9fbe-a30053b39b30)

**Comment by daltzctr** [comment]
![image](https://github.com/beto-rodriguez/LiveCharts2/assets/105223895/fa6fed5d-a202-4a1f-9909-b66ed6ff10cf)

**Comment by daltzctr** [comment]
Looks like `SolidColorPaint` is leaking anything that uses them.

**Comment by daltzctr** [comment]
Further analysis leads me to believe that https://github.com/beto-rodriguez/LiveCharts2/blob/60958aacd622793df898b341f8098a49009217ab/src/skiasharp/LiveChartsCore.SkiaSharp/Painting/Paint.cs#LL39C5-L39C114 is leaking somehow

**Comment by daltzctr** [comment]
It looks in `Measure()`

There is a unnecessary copy of the binded Series

```csharp
        var actualSeries = (_chartView.Series ?? Enumerable.Empty<ISeries>()).Where(x => x.IsVisible);

        Series = actualSeries
            .Cast<ICartesianSeries<TDrawingContext>>()
            .ToArray();
```

**Comment by daltzctr** [comment]
Is there symbols available for the public nuget that I could obtain for further testing? Pulling in source and testing it directly is painful due to the repository being outdated (VS2019, NET6.0, etc).

**Comment by daltzctr** [comment]
After even further debugging. I think the cause of the leak is right here

https://github.com/beto-rodriguez/LiveCharts2/blob/60958aacd622793df898b341f8098a49009217ab/src/skiasharp/LiveChartsCore.SkiaSharp/Painting/Paint.cs#L203-L213

geometries are being added to this HashSet without ever removing them.

This is called by 

https://github.com/beto-rodriguez/LiveCharts2/blob/60958aacd622793df898b341f8098a49009217ab/src/LiveChartsCore/VisualElements/RelativePanel.cs#L89

Which is then indirectly (via various measures propogating downwards) by `SKDefaultLegend.Measure()`. `SKDefaultLegend.Measure()` gets called by the main `CartesianChart.Measure()` every graph update (which in my case is constantly). 

This causes a memory leak of `SolidColorPaint` every 50ms (I'm updating the plot at 20ms, but throttled internally at 50ms). This leads the HashSet to continuously expand and any iterations upon it.

**Comment by daltzctr** [comment]
The following solves the memory leak for me

```
    private void Chart_UpdateFinished(LiveChartsCore.Kernel.Sketches.IChartView<LiveChartsCore.SkiaSharpView.Drawing.SkiaSharpDrawingContext> chart)
    {
        var coreCanvas = chart.CoreCanvas as MotionCanvas<SkiaSharpDrawingContext>;

        // Chart library leaks PaintTasks
        if (coreCanvas != null)
        {
            // Periodically clean up drawables, this may cause the chart to blip if the user mouses over it during this time
            if (coreCanvas.DrawablesCount > 50)
            {
                coreCanvas.SetPaintTasks(new HashSet<LiveChartsCore.Drawing.IPaint<SkiaSharpDrawingContext>>());
            }
        }
    }
    ```

**Comment by beto-rodriguez** [comment]
Thanks for the detailed research, I will take a look and come back.

**Comment by xwang-888** [comment]
> 感谢您的详细研究，我会看看并回来。

Has this issue been fixed in beta.802 ?

**Comment by xwang-888** [comment]
Will this issue only appear in beta 8.0?

**Comment by beto-rodriguez** [comment]
This issue is not fixed yet and is possible that also exists in previous versions.

**Comment by beto-rodriguez** [comment]
This is a critical issue that should be fixed in a future version.

can you provide a small repository or project with the issue? that will help a lot to fix this faster

**Comment by daltzctr** [comment]
@beto-rodriguez I haven't tested it, but your example project where you are constantly plotting data (make sure the visible axis are expanding and contracting) should reproduce it.

**Comment by xwang-888** [comment]
Like this. 

https://github.com/beto-rodriguez/LiveCharts2/assets/56998197/dfa83c40-cde0-4c1b-ba05-6fe78e2ca249

**Comment by xwang-888** [comment]
> 以下为我解决内存泄漏
> 
> ```
>     private void Chart_UpdateFinished(LiveChartsCore.Kernel.Sketches.IChartView<LiveChartsCore.SkiaSharpView.Drawing.SkiaSharpDrawingContext> chart)
>     {
>         var coreCanvas = chart.CoreCanvas as MotionCanvas<SkiaSharpDrawingContext>;
> 
>         // Chart library leaks PaintTasks
>         if (coreCanvas != null)
>         {
>             // Periodically clean up drawables, this may cause the chart to blip if the user mouses over it during this time
>             if (coreCanvas.DrawablesCount > 50)
>             {
>                 coreCanvas.SetPaintTasks(new HashSet<LiveChartsCore.Drawing.IPaint<SkiaSharpDrawingContext>>());
>             }
>         }
>     }
>     ```
> ```

Can adding this code temporarily solve the problem？

**Comment by ParkhutRoman** [comment]
It's leaks on page navigation too with empty CartesianChart, just with navigating back and forward I can see the same behavior.
![image](https://github.com/beto-rodriguez/LiveCharts2/assets/12161245/85099362-e1e0-4f3d-aed1-a157f50a35b5)

Version: v2.0.0-beta.710 (any newer have worse memory leaks)
.NET Version: .NET 7
Used in a WinUI3 project

**Comment by daltzctr** [comment]
Given direction on building this project for .NET 8, I wouldn't mind putting a small amount of effort into tracing the leak(s) and resolving them.

Last time I tried to pull it down and build it, it complained about lots of things being on .NET 6.

**Comment by daltzctr** [comment]
For anyone else, it'd help if when you see the memory leak, breakpoint and see how many drawables are queued up. I saw that as my main leak

**Comment by ParkhutRoman** [comment]
Graph must occupy new memory during creation, the problem is that it does not release all of it when unloading during navigation. No breakpoints will help here.

**Comment by daltzctr** [comment]
I don't see how breakpointing and seeing the amount of drawables queued up is not helping? Breakpoint and look at the CoreCanvas object and see if any of the contained structures contain a lot of data.

Regardless, it's likely your ...PointProperties are being held alive by SolidColorBrush. If you trace the roots you may find that to be the case.

**Comment by beto-rodriguez** [comment]
Also relevant: https://github.com/beto-rodriguez/LiveCharts2/issues/1083

**Comment by beto-rodriguez** [comment]
@ParkhutRoman I cannot reproduce the issue. Please provide a minimal reproducible repository where I can see the issue and fix that.

The memory leak does not happen in this repository, go to the run the WinUI sample in thsi repo,  then go to the Test/Dispose view, click multiple times time on the "change content" button:

![v](https://github.com/beto-rodriguez/LiveCharts2/assets/10853349/493217f1-c192-4fbc-a831-fe02c860fb0a)


No memory leaks visible!


Open the WinUI

**Comment by daltzctr** [comment]
@beto-rodriguez that GC pressure makes real time updating with multiple series really difficult.

**Comment by kshfire** [comment]
I also encounter similar problems.  Very simple use case.
  - Using   ObservableCollection<ObservablePoint>
  -             ScatterSeries
  - Using WPF's dispatcherTimer
  -          Every interval : update one point such as    
                                       Add(newValue)  and then  RemoveAt(0)

Then, By visual studio's diagnostic tools,  I found Memory usage is gradually and monotonically increased.
  - Using "Take Snapshot",  Memory is increasing...

Is it a real bug or my incorrect use??   
I want to update and show some data using dispatcher timer, but I found the memory leak issue.
I hope someone teach me to solve this issue.

**Comment by yangmijun** [comment]
Same problem:
       Using Prism mvvm framework. 
                Added 2 Views(and viewmodels), put some chart on the views.
                Registered the 2 views for navigation ( use RegisterForNavigation method in RegisterTypes)
                Every 10 secs, use RequestNavigate method to switch between the 2 views
                In the OnNavigatedTo  method of each viewmodel, try update the data, recreate the Series
Then I got almost same memory leaks as daltzctr.

**Comment by beto-rodriguez** [comment]
@kshfire That is a knows issue in WPF, WPF bindings can cause memory leaks if `INotifyPropertyChanged` is not implemented.

**Comment by beto-rodriguez** [comment]
There are multiple reasons that can cause a memory leak, **the sample in this repository do not have any memory leak**.

1. Please if you're facing a memory leak issue, ensure that the leak is not on your side.
2. Please provide a minimal reproducible sample, where I can see the issue (if exits)

**Comment by beto-rodriguez** [comment]
I think I see where the confusion is coming, in the sample on the website/repo I am doing:

https://github.com/beto-rodriguez/LiveCharts2/blob/master/samples/ViewModelsSamples/General/MultiThreading/ViewModel.cs

In the next block I call `Task.Run()`  to start a background task and test the library when the data comes from multiple threads.

That is just to keep the example simple, that example of course has a memory leak because I never stopped the `Task` I initiated with an infinite loop, but the leak is caused only in the sample, the chart has no leaks. I will add a note to that kind of samples to prevent future confusion.

Hopefully that case helps you to find the memory leaks in your cases, in the case of the sample, I could just stop the task I initiated when the view is disposed, then the GC will be able to do its job.
